Prepare docs and Swagger definition for automatically generated API docs

bfirsh opened the pull request:
**- What I did**

This is some preparation for automatically generating API documentation from a Swagger definition. See https://github.com/docker/docker/issues/22931.

It is the Engine half of [this documentation pull request](https://github.com/docker/docker.github.io/pull/606). This pull request should be merged before the documentation pull request.

It consists of a bunch of things:

- Removed the API reference for the latest API version. This is now all in `api/swagger.yaml`, which the documentation is automatically generated from.
- Removed all the other API documentation from this repository. The old versions barely ever change, and all the other stuff is not tied to any code change, so they should probably just live in docker/docker.github.io.
- Renamed "Remote API" to "Engine API". See https://github.com/docker/docker/issues/28319 for details.
- Update `api/swagger.yaml` with some more documentation and some recent API changes.

**- A picture of a cute animal (not mandatory but encouraged)**

![robot-cat](https://cloud.githubusercontent.com/assets/40906/20367010/8c6e2f6a-ac45-11e6-83dc-ac78a5e6e0b8.jpg)

/cc @dnephin @johndmulhausen

bfirsh committed:
Rename Remote API to Engine API

Implementation of https://github.com/docker/docker/issues/28319

Signed-off-by: Ben Firshman <ben@firshman.co.uk>

bfirsh committed:
Add version information to swagger.yaml

- Tidied up copy
- API version is now mandatory
- Links to previous API versions

Signed-off-by: Ben Firshman <ben@firshman.co.uk>

bfirsh committed:
Update swagger.yaml with changes to API

Up to and including 673c5ee559b05adb9ef6e0e1d6135ff83522e38d

Signed-off-by: Ben Firshman <ben@firshman.co.uk>

johndmulhausen commented:
So excited about this, thanks @bfirsh. Slowly, we are killing all these hand-rolled markdown files and I couldn't be happier. To be getting such nice-looking API docs is just an added bonus!

cc @mstanleyjones

thaJeztah commented:
Is there an easy way to generate these docs for preview / reviewing pull requests?

bfirsh commented:
@thaJeztah I think the easiest way is to copy `swagger.yaml` into the right place in the documentation repo and run the documentation locally.

That could be documented somewhere I suppose. Perhaps a "API documentation" section in the contributing docs?

dnephin commented on api/swagger.yaml:
```diff
@@ -6468,6 +6555,31 @@ paths:
+  /plugins/{name}/push:
+    post:
+      summary: "Push a plugin"
+      operationId: "PluginsPush"
```
Could we make this `PluginPush` ? I've been normalizing everything to be singular.

dnephin commented on api/swagger.yaml:
```diff
@@ -7067,6 +7246,84 @@ paths:
+        Get `stdout` and `stderr` logs from a service.
+
+        **Note**: This endpoint works only for services with the `json-file` or `journald` logging drivers.
+      operationId: "ServicesLogs"
```
`ServiceLogs`

dnephin commented on api/swagger.yaml:
```diff
@@ -7238,3 +7495,149 @@ paths:
+  /secrets:
+    get:
+      summary: "List secrets"
+      operationId: "SecretsList"
```
`SecretList`

dnephin commented on api/swagger.yaml:
```diff
@@ -7238,3 +7495,149 @@ paths:
+  /secrets/create:
+    post:
+      summary: "Create a secret"
+      operationId: "SecretsCreate"
```
`SecretCreate`

dnephin reviewed (commented):
Some minor requests for changes to OperationID.

Thanks for adding all the missing endpoints.

LGTM

bfirsh commented:
If we want this included in 1.13, then these changes need cherry-picking to the 1.13 branch. I see no reason why not to include it – it's purely a documentation change.

Following an offline discussion with @thaJeztah, I'm also going to add some docs about how to write and test the API docs.

thaJeztah commented:
Adding to the milestone, but /cc @vieux

bfirsh committed:
Add documentation for working on Engine API

Signed-off-by: Ben Firshman <ben@firshman.co.uk>

bfirsh commented:
@thaJeztah Documentation added! https://github.com/bfirsh/docker/blob/prepare-api-docs-for-swagger-docs/project/API.md

mdlinville reviewed (approved):
This is so cool!! Thank you so much, @bfirsh !

bfirsh committed:
Move API reference to new location

See: https://github.com/docker/docker.github.io/pull/606

Also:
- Add missing redirects to API reference pages
- Remove v1.25 and 1.26, because they are being replaced with
  swagger generated docs.
- Remove all other docs which aren't reference material, because
  this can live in docker/docker.github.io

Signed-off-by: Ben Firshman <ben@firshman.co.uk>

thaJeztah reviewed (approved):
LGTM

mlaventure commented on api/README.md:
```diff
@@ -1,5 +1,42 @@
+
+All the documentation generation is done in the documentation repository, [docker/docker.github.io](https://github.com/docker/docker.github.io). The Swagger definition is vendored periodically into this repository, but you can manually copy over the Swagger definition to test changes.
+
+Copy `api/swagger.yaml` in this repository to `engine/api/v1.25/swagger.yaml` in the documentation repository, overwriting what is already there. Then, run `docker-compose up` in the documentation repository and browse to [http://localhost:4000/engine/api/v1.25/](http://localhost:4000/engine/api/v1.25/) when it finishes rendering.
```
Replace `v1.25` with something like `<API_VERSION>` so we don't have to update this file every release? (we've already updated master to `v1.26`)

bfirsh replied:
Heh, good point. I'll just remove the version number actually, because `/engine/api/` redirects to `/engine/api/v1.25/`!

bfirsh replied:
@mlaventure Done!

vieux commented:
LGTM

thaJeztah commented:
There we go!

vieux commented:
@bfirsh can you create a PR against the 1.13.x branch directly ? this one is tricky to cherry-pick since it's 5 commits and the version changed in master

vieux commented:
@bfirsh nevermind I got it

bfirsh commented:
@vieux Nice, thank you! Appreciate that, because I realise it was non-trivial...

vieux commented:
@bfirsh https://github.com/docker/docker/blob/master/docs/api/version-history.md has 2 broken links, 1.25 and 1.26

bfirsh commented:
@vieux They will exist once https://github.com/docker/docker.github.io/pull/606 exists. :)

Well, 1.26 won't, but the `1.13` branch doesn't have that broken link.
